keras_contrib/applications/densenet.py

```diff
@@ -501,10 +527,24 @@ def __create_dense_net(nb_classes, img_input, include_top, depth=40, nb_dense_bl
     x = BatchNormalization(axis=concat_axis, gamma_regularizer=l2(weight_decay),
                            beta_regularizer=l2(weight_decay))(x)
     x = Activation('relu')(x)
-    x = GlobalAveragePooling2D()(x)
+    if pooling is not None:
+        x = GlobalAveragePooling2D()(x)
 
-    if include_top:
-        x = Dense(nb_classes, activation='softmax', kernel_regularizer=l2(weight_decay), bias_regularizer=l2(weight_decay))(x)
+    if include_top and pooling is not None:
```

Comment:
The goal is to support atrous convolutions, which will be a full image sized segmentation output (or a reduced one). There is no pooling in this case, and that is why the pooling option was added. In place of pooling there are atrous dilations to expand the receptive field, as if a 2x2 convolution covers the space of a 4x4 convolution when dilation_rate = 2.

I'll try to enumerate the important use cases:

 1. `pooling='max' || 'avg' && include_top=True || False` the final pooling layer is included, which is correct (look 2 lines above), perhaps pooling should be excluded when include_top is False? If this means moving line 530 and 531 to under line 533, that would seem reasonable. I was debating this before.
 2. `pooling=None && include_top=True` In this case the output will be 2D, even if dilation_rate=1, so the second 2D `top` is most appropriate, and GlobalAveragePooling would reduce the spatial resolution, which is undesirable for segmentation. 
 3. `pooling=None && include_top=False` Another 2D output case, pooling and top should both be entirely disabled, in this case the user will specify exactly what they want to do with the data, because it might be part of a completely different larger network.